include/kernels/gauge_polyakov_loop.cuh

```diff
@@ -0,0 +1,221 @@
+
+     @return The product of the gauge path
+     @param[in] arg Kernel argumnt
+     @param[in] x Full index array
+     @param[in] parity Parity index
+  */
+  template <int dir, typename Arg>
+  __device__ __host__ inline typename Arg::HighPrecLink
+  computePolyakovLoop(const Arg &arg, int x[4], int parity)
+  {
+    int dx[4] = {0, 0, 0, 0};
+
+    using Link = typename Arg::Link;
+    using HighPrecLink = typename Arg::HighPrecLink;
+
+    // polyloop: current matrix
+    // link: the loaded matrix in this round
+    Link link;
+    HighPrecLink hi_link, polyloop;
+    setIdentity(&polyloop);
+
+    int nbr_oddbit = parity;
+
+    for (int dt = 0; dt < arg.X[dir]; dt++) {
+      dx[dir] = dt;
+      link = arg.U(arg.geometry == QUDA_VECTOR_GEOMETRY ? dir : 0, linkIndexShift(x, dx, arg.X), nbr_oddbit);
+      hi_link = link; // promote
+      polyloop = polyloop * hi_link;
+      nbr_oddbit = nbr_oddbit ^ 1;
+    } // dt
+    return polyloop;
+  }
+
+  template <typename Float_, int nColor_, QudaReconstructType recon_>
+  struct GaugePolyakovLoopProductArg : public kernel_param<> {
+    using Float = Float_;
+    using AccumFloat = double;
+    static constexpr int nColor = nColor_;
+    static_assert(nColor == 3, "Only nColor=3 enabled at this time");
+    static constexpr QudaReconstructType recon = recon_;
+    static constexpr QudaFieldGeometry geometry = QUDA_VECTOR_GEOMETRY;
+    using Gauge = typename gauge_mapper<Float,recon>::type;
+    using AccumGauge = typename gauge_mapper<AccumFloat,recon>::type;
+    using Link = Matrix<complex<Float>, 3>;
+    using HighPrecLink = Matrix<complex<double>, 3>;
+
+    int X[4];
+    AccumGauge P;
+    Gauge U;
+
+    GaugePolyakovLoopProductArg(GaugeField &P_, const GaugeField &U_) :
+      kernel_param(dim3(P_.LocalVolumeCB(), 2, 1)),
+      P(P_),
+      U(U_)
+    {
+      if (U_.Geometry() != QUDA_VECTOR_GEOMETRY || P_.Geometry() != QUDA_SCALAR_GEOMETRY)
+        errorQuda("Unexpected geometry pair U %d ; P %d", U_.Geometry(), P_.Geometry());
+      for (int dir=0; dir<4; ++dir){
+        X[dir] = U_.X()[dir];
+        if (dir != 3 && U_.X()[dir] != P_.X()[dir]) errorQuda("Lengths %d %d in dimension %d do not agree", U_.X()[dir], P_.X()[dir], dir);
+      }
+    }
+
+  };
+
+  template <typename Arg> struct PolyakovLoopProduct {
+    const Arg &arg;
+    constexpr PolyakovLoopProduct(const Arg &arg) : arg(arg) {}
+    static constexpr const char *filename() { return KERNEL_FILE; }
+
+    __device__ __host__ inline void operator()(int x_cb, int parity)
+    {
+      int x[4];
+      getCoords(x, x_cb, arg.X, parity);
+
+      // hack
+      if (x[3] > 0) return;
+
+      auto polyloop = computePolyakovLoop<3>(arg, x, parity);
+
+      // save loop
+      arg.P(0, x_cb, parity) = polyloop;
+
+    }
+
+  };
+
+  template <typename Float_, int nColor_, QudaReconstructType recon_>
+  struct GaugePolyakovLoopTraceArg : public ReduceArg<array<double, 2>> {
+    using Float = Float_;
```

Comment:
As above: change this to `using real = typename mapper<Float>::type;`

Reply:
https://github.com/lattice/quda/commit/58845f78894fa8a26a831a6d0a61b8a50b56a269